Feat/user search

assokk opened the pull request:
## 작업 내용
- /users → /api/users 로 베이스 경로 통일
- 컨트롤러에 있던 일부 처리 로직을 서비스 계층으로 이관
- UserExternalService에서 타 도메인에게 유저 정보 전달 시 로직 개선

<br>

## 개선 사항 
- API 일관성 향상
- 레이어드 아키텍처 강화
- 공동개발 생산성 향상

<br>

## 개발(작업) 예정

- 

<br>

### 참고 사항

-

assokk committed:
feat: 유저 단건/다건 조회 서비스 구현 및 탈퇴 유저 제외 처리

assokk committed:
feat: 유저 조회 API 컨트롤러 구현

assokk committed:
refactor: User 조회 로직 및 레이어 구조 개선

assokk committed:
style: 사용하지 않는 import문 제거

donghyeon505 commented on src/main/java/com/example/taskflow/domain/user/controller/UserController.java:
```diff
@@ -1,75 +1,41 @@
-
-        Optional<User> optionalUser = userInternalService.findByIdAndDeletedAtIsNull(userId);
+    public ResponseEntity<ApiResponse<UserResponse>> getMyInfo(
+            @RequestAttribute(value = "userId", required = false) Long userId) {
```
자신의 정보를 조회하기 때문에 required 가 false 가 되면 안될 것 같습니다!

donghyeon505 commented on src/main/java/com/example/taskflow/domain/user/service/UserExternalService.java:
```diff
@@ -1,38 +1,50 @@
-    public Optional<User> findByIdAndDeletedAtIsNull(Long id) {
-        return userRepository.findByIdAndDeletedAtIsNull(id);
+    // 1. 단일 유저 조회 (탈퇴한 유저 제외)
+    public UserResponse getFindById(Long id) {
```
UserResponse 가 아닌 User 객체를 전달 해주셔야합니다!

donghyeon505 commented on src/main/java/com/example/taskflow/domain/user/service/UserExternalService.java:
```diff
@@ -1,38 +1,50 @@
-                .filter(user -> user.getDeletedAt() == null)
+    // 2. 여러 유저 조회 (탈퇴한 유저 제외)
+    // 팀 멤버 목록 조회 시 사용
+    public List<UserResponse> getActiveMemberByTeamId(Long teamId) {
```
다른 도메인의 Repository를 직접 조회하지 않기 위해서 ExternalService를 쓰기 때문에 직접 조회하는 것이 아닌,
TeamId를 사용해서 User를 조회해서 List<User> 를 던져 주어야 한다고 봅니다!

donghyeon505 commented on src/main/java/com/example/taskflow/domain/user/service/UserInternalService.java:
```diff
@@ -1,30 +1,52 @@
+    public List<UserResponse> getAvailableUsers(Long teamId) {
+
+        if (!teamRepository.existsById(teamId)) {
+            throw new GlobalException(UserErrorCode.TEAM_NOT_FOUND); }
```
이 부분은 TeamExternalService의 책임 이라고 봅니다.

assokk committed:
refactor: 사용자 조회 기능 전반 리팩토링

- teamRepository 직접 접근 제거
- 타 도메인에서 조회 메서드 사용 시 반환 타입을 User 객체로 변경
- TeamExternalService로 책임 분할
- enum 대문자 통일

assokk committed:
refactor: 사용하지 않는 import 제거 및 enum 정리

donghyeon505 commented on src/main/java/com/example/taskflow/domain/user/controller/UserController.java:
```diff
@@ -35,6 +35,8 @@ public ResponseEntity<ApiResponse<UserResponse>> getMyInfo(
     @GetMapping("/available")
     public ResponseEntity<ApiResponse<List<UserResponse>>> getAvailableUsers(@RequestParam("teamId") Long teamId) {
 
```
사용자 목록 조회할 때, API 명세서를 보니 teamId 가 선택 사항이며, 있을 시에만 해당 팀에 속하지 않은 사용자 목록을 조회해야 할 것 같습니다! required = false 로 받아야할 것 같습니다!

donghyeon505 commented on src/main/java/com/example/taskflow/domain/user/service/UserInternalService.java:
```diff
@@ -38,9 +37,6 @@ public UserResponse getMyInfo(Long userId) {
     // 2. 팀에 추가 가능한 사용자 목록 조회
     public List<UserResponse> getAvailableUsers(Long teamId) {
 
```
마찬가지로 teamId 가 없을 때의 리턴(findAll)도 줄 수 있도록 해야할 것 같습니다!

assokk replied:
넵!!! 감사합니다

assokk committed:
refactor: 팀에 추가 가능한 사용자 목록 조회 기능 개선 및 메서드 명 수정

- teamId 미전달 시 전체 사용자 조회되도록 로직 분기 처리
- 레포지토리 및 서비스 메서드명을 기능에 맞게 명확하게 변경

assokk committed:
feat: deletedAt 세터 추가로 auth 탈퇴 로직 연동

dbdnstjs reviewed (approved):
수고하셨습니다

Hyjpark commented on src/main/java/com/example/taskflow/domain/user/controller/UserController.java:
```diff
@@ -1,74 +1,43 @@
-
-        Optional<User> optionalUser = userInternalService.findByIdAndDeletedAtIsNull(userId);
+    public ResponseEntity<ApiResponse<UserResponse>> getMyInfo(
+            @RequestAttribute(value = "userId") Long userId) {
```
사용하는데 이슈가 있지 않으신다면  `@RequestAttribute(value = "userId") Long userId` 부분은
커스텀 어노테이션을 지원하고 있으므로
`@CurrentUser Long userId`로 바꿔서 쓰시면 좋을 거 같습니다. 
기능에 문제가 있는 부분은 아니지만 다른 도메인과의 통일성을 위해 부탁드립니다!

assokk replied:
넵!!!

Hyjpark commented on src/main/java/com/example/taskflow/domain/user/entity/User.java:
```diff
@@ -36,11 +37,19 @@ public class User extends BaseEntity {
 
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "team_id")
+    private Team team;
```
이부분은 이번에 TeamMermber 엔티티가 추가되면 User와 Team 간의 매핑이 필요하지 않게 되었습니다!
User가 team.id를 가지게 될 겨우 하나의 팀에만 소속될 수 있는 구조가 되므로
팀이 없는 유저를 조회할 경우 TeamMermber를 통해 처리하는 방향으로 리팩토링 하시면 될 것 같습니다.
자세한 내용은 회의 노션을 참고해주세요

assokk replied:
넵!!!

Hyjpark commented on src/main/java/com/example/taskflow/domain/user/service/UserInternalService.java:
```diff
@@ -1,30 +1,71 @@
+        List<User> users = userRepository.findUsersNotInTeam(teamId);
+
+        List<UserResponse> responseList = new ArrayList<>();
+        for (User user : users) {
```
stream + map으로 리팩토링하면 코드가 더 깔끔해집니다. 리팩토링하실때 참고하시면 좋겠습니다.

assokk replied:
넵!!!

Hyjpark commented on src/main/java/com/example/taskflow/domain/user/repository/UserRepository.java:
```diff
@@ -12,16 +12,18 @@ public interface UserRepository extends JpaRepository<User, Long> {
-          )
-        """)
-    List<User> findAvailableUsersByTeamId(Long teamId);
+            AND (u.team.id != :teamId OR u.team IS NULL)
```
위의 TeamMember에 관한 리뷰 참고하셔서 리팩토링하시면 될 거 같습니다!

assokk replied:
넵 감사합니다 :D

assokk committed:
feat: Search 도메인 DTO 생성

assokk commented:
최신 dev 가져와서 마지막 커밋 올린 상태라 아마 깃 충돌 문제는 해결되었을 것 같습니다!
승인 주시면 해결되었는지 확실하게 확인해볼 수 있을 것 같습니다.
또한 혜정님이 피드백 주신 부분 수정 중입니다!
수정 후 말씀드리겠습니다 그때 검토 부탁드리겠습니다 :D

assokk committed:
feat: 통합 검색 기능 구현

- task, user, team에 대한 키워드 검색 기능 구현

assokk committed:
feat: 유저 도메인 전용 예외 처리 코드 추가

assokk committed:
refactor: 서비스 로직 stream+map 리팩토링 및 UserResponse DTO 정적 팩토리 메서드 적용

assokk committed:
refactor: @RequestAttribute → @CurrentUser 어노테이션 적용으로 코드 일관성 개선

assokk committed:
refactor: User-Team 연관관계 제거 및 TeamMember 일대다 매핑으로 리팩토링

assokk committed:
feat: Search 도메인에서 사용할 사용자 검색 기능 추가

assokk committed:
refactor: 주석 정리

Kangsunggyu reviewed (approved):
external 부분과 enum 부분 확인했습니다. 수고하셨습니다.
